// @percy/config/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "addMigration", {
  enumerable: true,
  get: function () {
    return _migrate.addMigration;
  }
});
Object.defineProperty(exports, "addSchema", {
  enumerable: true,
  get: function () {
    return _validate.addSchema;
  }
});
Object.defineProperty(exports, "cache", {
  enumerable: true,
  get: function () {
    return _load.cache;
  }
});
Object.defineProperty(exports, "clearMigrations", {
  enumerable: true,
  get: function () {
    return _migrate.clearMigrations;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "explorer", {
  enumerable: true,
  get: function () {
    return _load.explorer;
  }
});
Object.defineProperty(exports, "getDefaults", {
  enumerable: true,
  get: function () {
    return _defaults.default;
  }
});
Object.defineProperty(exports, "load", {
  enumerable: true,
  get: function () {
    return _load.default;
  }
});
Object.defineProperty(exports, "migrate", {
  enumerable: true,
  get: function () {
    return _migrate.default;
  }
});
Object.defineProperty(exports, "normalize", {
  enumerable: true,
  get: function () {
    return _normalize.default;
  }
});
Object.defineProperty(exports, "resetSchema", {
  enumerable: true,
  get: function () {
    return _validate.resetSchema;
  }
});
Object.defineProperty(exports, "search", {
  enumerable: true,
  get: function () {
    return _load.search;
  }
});
Object.defineProperty(exports, "stringify", {
  enumerable: true,
  get: function () {
    return _stringify.default;
  }
});
Object.defineProperty(exports, "validate", {
  enumerable: true,
  get: function () {
    return _validate.default;
  }
});

var _load = _interopRequireWildcard(require("./load"));

var _validate = _interopRequireWildcard(require("./validate"));

var _migrate = _interopRequireWildcard(require("./migrate"));

var _defaults = _interopRequireDefault(require("./defaults"));

var _normalize = _interopRequireDefault(require("./normalize"));

var _stringify = _interopRequireDefault(require("./stringify"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// mirror the namespace as the default export
var _default = {
  load: _load.default,
  search: _load.search,
  cache: _load.cache,
  explorer: _load.explorer,
  validate: _validate.default,
  addSchema: _validate.addSchema,
  resetSchema: _validate.resetSchema,
  migrate: _migrate.default,
  addMigration: _migrate.addMigration,
  clearMigrations: _migrate.clearMigrations,
  getDefaults: _defaults.default,
  normalize: _normalize.default,
  stringify: _stringify.default
};
exports.default = _default;
