// @percy/cli-doctor/dist/utils/http.js
function _classPrivateMethodInitSpec(e, a) { _checkPrivateRedeclaration(e, a), a.add(e); }
function _checkPrivateRedeclaration(e, t) { if (t.has(e)) throw new TypeError("Cannot initialize the same private elements twice on an object"); }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _assertClassBrand(e, t, n) { if ("function" == typeof e ? e === t : e.has(t)) return arguments.length < 3 ? t : n; throw new TypeError("Private element is not present on this object"); }
import https from 'https';
import http from 'http';
import net from 'net';
import tls from 'tls';
import { URL } from 'url';

// Default request timeout (10 seconds)
const DEFAULT_TIMEOUT = 10000;

/**
 * Probe result object
 * @typedef {object} ProbeResult
 * @property {boolean} ok        - True if request succeeded (2xx/3xx)
 * @property {number}  status    - HTTP status code (0 if no response)
 * @property {string}  error     - Error message if request failed
 * @property {string}  errorCode - Node.js error code (e.g. CERT_HAS_EXPIRED)
 * @property {number}  latencyMs - Round-trip time in milliseconds
 * @property {string}  body      - Response body as UTF-8 string
 */
var _HttpProber_brand = /*#__PURE__*/new WeakSet();
export class HttpProber {
  constructor() {
    /**
     * Unified request dispatcher.  Routes to the right transport based on whether
    * a proxy is present and whether the target URL uses HTTPS.
    *
    *  • no proxy               → direct HTTP or HTTPS
    *  • proxy + HTTPS target   → CONNECT tunnel → TLS → HTTPS request
    *  • proxy + HTTP target    → plain HTTP request with absolute-URI to proxy
    */
    _classPrivateMethodInitSpec(this, _HttpProber_brand);
  }
  /**
   * Make a probe HTTP/HTTPS request, resolving with a ProbeResult rather than
   * throwing so callers can inspect what went wrong.
   *
   * @param {string} targetUrl
   * @param {object} [options]
   * @param {string}  [options.proxyUrl]  - http(s)://host:port proxy
   * @param {number}  [options.timeout]   - ms before abort (default 10 000)
   * @param {string}  [options.method]    - HTTP verb (default HEAD)
   * @returns {Promise<ProbeResult>}
   */
  async probeUrl(targetUrl, options = {}) {
    const {
      proxyUrl,
      timeout = DEFAULT_TIMEOUT,
      method = 'HEAD',
      headers = {}
    } = options;
    const start = Date.now();
    try {
      const url = new URL(targetUrl);
      const proxy = proxyUrl ? new URL(proxyUrl) : null;
      const result = await _assertClassBrand(_HttpProber_brand, this, _makeRequest).call(this, url, proxy, {
        timeout,
        method,
        headers
      });
      result.latencyMs = Date.now() - start;
      return result;
    } catch (err) {
      return {
        ok: false,
        status: 0,
        error: err.message,
        errorCode: /* istanbul ignore next */err.code ?? 'UNKNOWN',
        latencyMs: Date.now() - start
      };
    }
  }

  /**
   * Returns true when the probe result carries an SSL/TLS error code.
   *
   * @param {ProbeResult} result
   * @returns {boolean}
   */
  isSslError(result) {
    if (!result || !result.errorCode) return false;
    if (HttpProber.SSL_ERROR_CODES.has(result.errorCode)) return true;
    return result.errorCode.includes('CERT') || result.errorCode.includes('SSL');
  }
}
function _makeRequest(url, proxy, {
  timeout,
  method,
  headers
}) {
  // ── common response resolver ──────────────────────────────────────────────
  function resolveResponse(res, resolve) {
    const status = res.statusCode;
    const chunks = [];
    res.on('data', chunk => chunks.push(chunk));
    res.on('end', () => {
      const body = Buffer.concat(chunks).toString('utf8');
      resolve({
        ok: status >= 200 && status < 400,
        status,
        error: null,
        errorCode: null,
        responseHeaders: res.headers,
        body
      });
    });
  }

  // ── common timeout error ──────────────────────────────────────────────────
  function timeoutError(label) {
    return Object.assign(new Error(`${label} timed out after ${timeout}ms`), {
      code: 'ETIMEDOUT'
    });
  }

  // ─────────────────────────────────────────────────────────────────────────
  // (A) HTTPS via CONNECT proxy tunnel
  // ─────────────────────────────────────────────────────────────────────────
  if (proxy && url.protocol === 'https:') {
    return new Promise((resolve, reject) => {
      const proxyPort = parseInt(proxy.port, 10) || (proxy.protocol === 'https:' ? 443 : 8080);
      const targetPort = parseInt(url.port, 10) || 443;
      const socket = net.connect({
        host: proxy.hostname,
        port: proxyPort
      });
      let tlsSock = null;
      let settled = false;
      const fail = err => {
        /* istanbul ignore next */
        if (settled) return;
        settled = true;
        clearTimeout(timer);
        try {
          var _tlsSock;
          (_tlsSock = tlsSock) === null || _tlsSock === void 0 || _tlsSock.destroy();
        } catch {/* ignore */}
        try {
          socket.destroy();
        } catch {/* ignore */}
        reject(err);
      };

      /* istanbul ignore next */
      const pass = result => {
        /* istanbul ignore next */
        if (settled) return;
        settled = true;
        clearTimeout(timer);
        resolve(result);
      };
      const timer = setTimeout(() => {
        fail(timeoutError('Proxy CONNECT'));
      }, timeout);
      socket.on('error', fail);
      socket.on('connect', () => {
        const authHeader = proxy.username ? `Proxy-Authorization: Basic ${Buffer.from(`${proxy.username}:${proxy.password}`).toString('base64')}\r\n` : '';
        socket.write(`CONNECT ${url.hostname}:${targetPort} HTTP/1.1\r\n` + `Host: ${url.hostname}:${targetPort}\r\n` + authHeader + '\r\n');
        let response = '';
        socket.on('data', chunk => {
          response += chunk.toString();
          if (!response.includes('\r\n\r\n')) return;
          const statusLine = response.split('\r\n')[0];
          const statusCode = parseInt(statusLine.split(' ')[1], 10);
          if (statusCode !== 200) {
            return fail(Object.assign(new Error(`Proxy CONNECT failed: ${statusLine}`), {
              code: 'EPROXY'
            }));
          }
          socket.removeAllListeners('data');
          tlsSock = tls.connect({
            socket,
            servername: url.hostname
          });
          tlsSock.on('error', fail);
          /* istanbul ignore next */
          tlsSock.on('secureConnect', () => {
            const req = https.request({
              createConnection: () => tlsSock,
              hostname: url.hostname,
              port: targetPort,
              path: url.pathname + url.search,
              method,
              headers
            }, /* istanbul ignore next */
            res => {
              const status = res.statusCode;
              const chunks = [];
              res.on('data', chunk => chunks.push(chunk));
              res.on('end', () => {
                const body = Buffer.concat(chunks).toString('utf8');
                pass({
                  ok: status >= 200 && status < 400,
                  status,
                  error: null,
                  errorCode: null,
                  responseHeaders: res.headers,
                  body
                });
              });
            });
            req.on('error', fail);
            req.end();
          });
        });
      });
    });
  }

  // ─────────────────────────────────────────────────────────────────────────
  // (B) HTTP via proxy (plain absolute-URI request, no tunnelling)
  // ─────────────────────────────────────────────────────────────────────────
  if (proxy && url.protocol === 'http:') {
    return new Promise((resolve, reject) => {
      const proxyPort = parseInt(proxy.port, 10) || 8080;
      const req = http.request({
        hostname: proxy.hostname,
        port: proxyPort,
        path: url.href,
        // absolute URI
        method,
        timeout,
        headers: {
          Host: url.hostname,
          ...headers,
          ...(proxy.username ? {
            'Proxy-Authorization': `Basic ${Buffer.from(`${proxy.username}:${proxy.password}`).toString('base64')}`
          } : {})
        }
      }, res => {
        if (res.statusCode === 407) {
          res.resume();
          return reject(Object.assign(new Error('Proxy CONNECT failed: HTTP/1.1 407 Proxy Authentication Required'), {
            code: 'EPROXY'
          }));
        }
        resolveResponse(res, resolve);
      });
      req.on('timeout', () => {
        req.destroy();
        reject(timeoutError('Request'));
      });
      req.on('error', reject);
      req.end();
    });
  }

  // ─────────────────────────────────────────────────────────────────────────
  // (C) Direct HTTP or HTTPS (no proxy)
  // ─────────────────────────────────────────────────────────────────────────
  return new Promise((resolve, reject) => {
    const driver = url.protocol === 'https:' ? https : http;
    const port = url.port || (url.protocol === 'https:' ? 443 : 80);
    const req = driver.request({
      hostname: url.hostname,
      port,
      path: url.pathname + url.search,
      method,
      timeout,
      headers
    }, res => resolveResponse(res, resolve));
    req.on('timeout', () => {
      req.destroy();
      reject(timeoutError('Request'));
    });
    req.on('error', reject);
    req.end();
  });
}
// SSL-related error codes emitted by Node's TLS/crypto layer
_defineProperty(HttpProber, "SSL_ERROR_CODES", new Set(['CERT_HAS_EXPIRED', 'CERT_NOT_YET_VALID', 'CERT_REVOKED', 'DEPTH_ZERO_SELF_SIGNED_CERT', 'ERR_TLS_CERT_ALTNAME_INVALID', 'SELF_SIGNED_CERT_IN_CHAIN', 'UNABLE_TO_GET_ISSUER_CERT', 'UNABLE_TO_VERIFY_LEAF_SIGNATURE', 'UNABLE_TO_GET_ISSUER_CERT_LOCALLY']));
export const httpProber = new HttpProber();
