// @percy/cli-command/dist/flags.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.verbose = exports.silent = exports.quiet = exports.networkIdleTimeout = exports.dryRun = exports.disableCache = exports.debug = exports.config = exports.allowedHostnames = exports.PERCY = exports.GLOBAL = exports.DISCOVERY = void 0;

var _logger = _interopRequireDefault(require("@percy/logger"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Global CLI flags
const verbose = {
  name: 'verbose',
  description: 'Log everything',
  exclusive: ['quiet', 'silent'],
  parse: () => !!_logger.default.loglevel('debug'),
  group: 'Global',
  short: 'v'
};
exports.verbose = verbose;
const quiet = {
  name: 'quiet',
  description: 'Log errors only',
  exclusive: ['verbose', 'silent'],
  parse: () => !!_logger.default.loglevel('warn'),
  group: 'Global',
  short: 'q'
};
exports.quiet = quiet;
const silent = {
  name: 'silent',
  description: 'Log nothing',
  exclusive: ['verbose', 'quiet'],
  parse: () => !!_logger.default.loglevel('silent'),
  group: 'Global',
  short: 's'
}; // Common percy and asset discovery flags

exports.silent = silent;
const config = {
  name: 'config',
  description: 'Config file path',
  percyrc: 'config',
  type: 'file',
  group: 'Percy',
  short: 'c'
};
exports.config = config;
const dryRun = {
  name: 'dry-run',
  description: 'Print snapshot names only',
  percyrc: 'dryRun',
  group: 'Percy',
  short: 'd'
};
exports.dryRun = dryRun;
const allowedHostnames = {
  name: 'allowed-hostname',
  description: 'Allowed hostnames to capture in asset discovery',
  percyrc: 'discovery.allowedHostnames',
  type: 'hostname',
  multiple: true,
  group: 'Percy',
  short: 'h'
};
exports.allowedHostnames = allowedHostnames;
const networkIdleTimeout = {
  name: 'network-idle-timeout',
  description: 'Asset discovery network idle timeout',
  percyrc: 'discovery.networkIdleTimeout',
  type: 'ms',
  parse: Number,
  group: 'Percy',
  short: 't'
};
exports.networkIdleTimeout = networkIdleTimeout;
const disableCache = {
  name: 'disable-cache',
  description: 'Disable asset discovery caches',
  percyrc: 'discovery.disableCache',
  group: 'Percy'
};
exports.disableCache = disableCache;
const debug = {
  name: 'debug',
  description: 'Debug asset discovery and do not upload snapshots',
  parse: () => !!_logger.default.loglevel('debug'),
  percyrc: 'skipUploads',
  group: 'Percy'
}; // Group constants

exports.debug = debug;
const GLOBAL = [verbose, quiet, silent];
exports.GLOBAL = GLOBAL;
const PERCY = [config, dryRun];
exports.PERCY = PERCY;
const DISCOVERY = [allowedHostnames, networkIdleTimeout, disableCache, debug];
exports.DISCOVERY = DISCOVERY;
